From: Thomas Renninger <trenn@suse.de>
Subject: PERF: Fix perf timechart C-state regression
References: none
Patch-Mainline: Submitted

Could not go into stable as it is integrated into a bigger
patch set which is still pending for mainline inclusion.

The bug got introduced with commit 4c21adf26f8fcf86a755b9b9f5.

The variable pe (struct power_entry) must only be used by
power:power_start event:
struct power_entry {
	struct trace_entry te;
	u64	type;
	u64	value;
	u64	cpu_id;
};

Doing this 100% correctly a new struct has to be defined like:
struct power_end_entry {
	struct trace_entry te;
	u64	cpu_id;
};

Then C-state changes initiated from a foreign core can be tracked
as well. But HW being able to do this does not exist and power_end
event gets deprecated soon anyway.

Signed-off-by: Thomas Renninger <trenn@suse.de>
CC: Arjan van de Ven <arjan@linux.intel.com>
CC: Ingo Molnar <mingo@elte.hu>
CC: stable@kernel.org


diff --git a/tools/perf/builtin-timechart.c b/tools/perf/builtin-timechart.c
index 9bcc38f..b3028eb 100644
--- a/tools/perf/builtin-timechart.c
+++ b/tools/perf/builtin-timechart.c
@@ -502,7 +502,7 @@ static int process_sample_event(event_t *event, struct perf_session *session)
 			c_state_start(pe->cpu_id, data.time, pe->value);
 
 		if (strcmp(event_str, "power:power_end") == 0)
-			c_state_end(pe->cpu_id, data.time);
+			c_state_end(data.cpu, data.time);
 
 		if (strcmp(event_str, "power:power_frequency") == 0)
 			p_state_change(pe->cpu_id, data.time, pe->value);
